### PR TITLE
Initialize cloud provider in node

### PR DESCRIPTION
**Pull request by jsafrane**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1379600

Be careful with review, I don't know these parts of OpenShift very well. I tested AWS and GCE and it fixes the bug there.

### CODE REVIEW

**Comment by jsafrane** [comment]
[test]

File: pkg/cmd/server/kubernetes/node_config.go [hunk 1]
```diff
@@ -214,6 +216,18 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig, enableProxy, enable
...
 	}
 
+	// Initialize cloud provider
+	if len(server.CloudProvider) > 0 && server.CloudProvider != v1alpha1.AutoDetectCloudProvider {
+		cloud, err := cloudprovider.InitCloudProvider(server.CloudProvider, server.CloudConfigFile)
+		if err != nil {
+			return nil, err
+		}
+		if cloud != nil {
+			glog.V(2).Infof("Successfully initialized cloud provider: %q from the config file: %q\n", server.CloudProvider, server.CloudConfigFile)
```
Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config.go]: No need for `\n`

**Comment by ncdc** [comment]
cc @liggitt @deads2k @pmorie @derekwaynecarr @smarterclayton @mfojtik

File: pkg/cmd/server/kubernetes/node_config.go (hunk 1 above)
Comment (jsafrane) [inline pkg/cmd/server/kubernetes/node_config.go]: fixed, however master prints with '\n': https://github.com/openshift/origin/blob/master/pkg/cmd/server/kubernetes/master_config.go#L213

**Comment by jsafrane** [comment]
> Be careful with review, I don't know these parts of OpenShift very well.

I mean, it does the right thing, I am not sure it does it on the right place.

File: pkg/cmd/server/kubernetes/node_config.go [hunk 2]
```diff
@@ -214,6 +216,18 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig, enableProxy, enable
 		return nil, err
 	}
 
+	// Initialize cloud provider
+	if len(server.CloudProvider) > 0 && server.CloudProvider != v1alpha1.AutoDetectCloudProvider {
```
Comment (jsafrane) [inline pkg/cmd/server/kubernetes/node_config.go]: I added some simple unit tests for server.CloudProvider parsing, however there is not a single unit test that would test whole `BuildKubernetesNodeConfig`. This function does many decisions that are not tested anywhere and building a cloud provider is just one of its duties.

IMO, there should be e2e tests that would catch GCE or AWS volume plugin throwing errors when mounting a volume.

File: pkg/cmd/server/kubernetes/node_config_test.go [hunk 3]
```diff
@@ -157,3 +160,49 @@ func TestProxyConfig(t *testing.T) {
...
+func TestBuildCloudProviderFake(t *testing.T) {
+	providerName := "fake"
+	cloudprovider.RegisterCloudProvider(providerName, func(config io.Reader) (cloudprovider.Interface, error) {
+		return &fake.FakeCloud{}, nil
+	})
+
+	server := kubeletoptions.NewKubeletServer()
+	server.CloudProvider = providerName
+
+	cloud, err := buildCloudProvider(server)
```
Comment (ncdc) [inline pkg/cmd/server/kubernetes/node_config_test.go]: @smarterclayton are you looking for a test that ensures that BuildKubernetesNodeConfig ends up with the cloud provider filled in correctly?

**Comment by smarterclayton** [comment]
Yes.

**Comment by ncdc** [comment]
@jsafrane do you think you could take a stab at writing a new test that calls BuildKubernetesNodeConfig?

**Comment by jsafrane** [comment]
@ncdc no, I really do not feel comfortable testing this code. I've barely seen it and it looks pretty complex to me.

**Comment by ncdc** [comment]
I'll write one

**Comment by jsafrane** [comment]
Thanks! I'd like to point out that this PR blocks testing of OpenShift on GCE and AWS, our QE literally  cannot do anything there and there is plenty of new stuff to test in 1.4.

**Comment by ncdc** [comment]
Let's merge this as-is and I'll do a follow-up with the test (assuming it's not impossible).

LGTM. Let's get 1 more +1 and then we'll tag - @liggitt @derekwaynecarr @mfojtik @pweil- @deads2k anyone? 😄

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**